u2f-ref-code/java/src/com/google/u2f/codec/RawMessageCodec.java

```diff
@@ -192,6 +194,80 @@ public static AuthenticateResponse decodeAuthenticateResponse(byte[] data) throw
     }
   }
 
+  public static TransferAccessResponse decodeTransferAccessResponse(byte[] data) 
+      throws U2FException {
+    final int SEQUENCE_NUMBER_SIZE = 1;
+    final int PUBLIC_KEY_SIZE = 65;
+    final int APP_ID_SIZE = 32;
+    
+	try {
+	  DataInputStream inputStream = new DataInputStream(new ByteArrayInputStream(data));
+	  byte controlByte = inputStream.readByte();
+	  
+	  if (!inputStream.markSupported()) {
+	    throw new U2FException("Stream doesn't support mark and reset.");
+	  }
+	  inputStream.mark(inputStream.available());
+      int numberOfTransferAccessMessages = inputStream.readUnsignedByte();
+      inputStream.reset();
+
+      TransferAccessMessage[] transferAccessMessages =
+          new TransferAccessMessage[numberOfTransferAccessMessages];    
+      
+      for (int i = 0; i < numberOfTransferAccessMessages; i ++ ) {
+        inputStream.mark(inputStream.available());
+        inputStream.skipBytes(SEQUENCE_NUMBER_SIZE + PUBLIC_KEY_SIZE + APP_ID_SIZE);
+        
+        X509Certificate newAttestationCertificate = (X509Certificate) CertificateFactory
+            .getInstance("X.509").generateCertificate(inputStream);
+        byte[] newAttestationCertificateBytes;
+        try {
+          newAttestationCertificateBytes = newAttestationCertificate.getEncoded();
+        } catch (CertificateEncodingException e) {
+          throw new U2FException("Error when encoding attestation certificate.", e);
+        }
+
+        int lengthOfAttestationCertificate = newAttestationCertificateBytes.length;
+
+        int lengthOfSignatureUsingPrivateKey = inputStream.readUnsignedByte();
+        inputStream.skipBytes(lengthOfSignatureUsingPrivateKey);
+        int lengthOfSignatureUsingAttestationKey = inputStream.readUnsignedByte();
+        inputStream.reset();
+
+        int lengthOfRawTransferAccessMessage =
+            SEQUENCE_NUMBER_SIZE + PUBLIC_KEY_SIZE + APP_ID_SIZE + 1 + lengthOfAttestationCertificate
+                + 1 + lengthOfSignatureUsingPrivateKey + lengthOfSignatureUsingAttestationKey;
+
+        byte[] rawTransferAccessMessage = new byte[lengthOfRawTransferAccessMessage];
+        inputStream.readFully(rawTransferAccessMessage);
+        transferAccessMessages[numberOfTransferAccessMessages - i - 1] =
+            TransferAccessMessage.fromBytes(rawTransferAccessMessage);
+        if (transferAccessMessages[numberOfTransferAccessMessages - i - 1]
+            .getMessageSequenceNumber() != numberOfTransferAccessMessages - i) {
+          throw new U2FException("Messages not in order, unexpected sequence number");
```

Comment:
The first thing that comes to mind is to use a string builder and formatter to build the string, and then call 
throw new U2fException(builtStringBuilder.toString());

does this seem acceptable?